### PR TITLE
feat-#3: Made the entire backend code from javascript to typescript

### LINKED ISSUE #3: Refactoring backend using typescript

### PR DESCRIPTION
**Pull request by Ameerjafar**
## Summary
Made the entire backend code from javascript to typescript.
## Description

https://github.com/user-attachments/assets/a949160c-26b1-44e6-9508-d18ec62954ad

currently in the project the backend code has not written by the typescript This pr converted the entire backend javascript to typescript.

## Issue(s) Addressed

_Enter the issue number of the bug(s) that this PR fixes_

- Template should be strictly **Closes #3 **

## Prerequisites

- [ Yes ] Have you followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions)?

### CODE REVIEW

**Comment by Ameerjafar** [comment]
@krishnaacharyaa can you review the pr sir

**Comment by Ameerjafar** [comment]
@krishnaacharyaa

File: backend/controllers/posts-controller.ts [hunk 1]
```diff
@@ -49,41 +49,41 @@ export const createPostHandler = async (req, res) => {
 
     const [savedPost] = await Promise.all([
       post.save(), // Save the post
-      deleteDataFromCache(REDIS_KEYS.ALL_POSTS), // Invalidate cache for all posts
-      deleteDataFromCache(REDIS_KEYS.FEATURED_POSTS), // Invalidate cache for featured posts
-      deleteDataFromCache(REDIS_KEYS.LATEST_POSTS), // Invalidate cache for latest posts
+      // deleteDataFromCache(REDIS_KEYS.ALL_POSTS), // Invalidate cache for all posts
```
Comment (krishnaacharyaa) [inline backend/controllers/posts-controller.ts:52]: What is the problem basically we are facing here?

File: backend/package.json [hunk 2]
```diff
@@ -1,6 +1,19 @@
 {
-  "type": "module",
+  "type": "commonjs",
   "dependencies": {
```
Comment (krishnaacharyaa) [inline backend/package.json:3]: Why this change?

File: backend/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,108 @@
+{
```
Comment (krishnaacharyaa) [inline backend/tsconfig.json:1]: Can we clean this up please

**Review by krishnaacharyaa — COMMENTED** [review]
Thank you for taking time and refactoring the code to ts, much appreciated, 
Kindly address the review comments, looks good to me 

p.s: Sorry didn't get time to review the PR because of office work

File: backend/package.json (hunk 2 above)
Comment (Ameerjafar) [reply backend/package.json:3]: In the type module we cannot use the exports or module.exports(we have used this this is somewhere in our codebase i have checked but i don't know where we have used that. this is the reason i have useed type="commonjs" @krishnaacharyaa

File: backend/tsconfig.json (hunk 3 above)
Comment (Ameerjafar) [reply backend/tsconfig.json:1]: Can I remove the full command code ?

File: backend/controllers/posts-controller.ts (hunk 1 above)
Comment (Ameerjafar) [reply backend/controllers/posts-controller.ts:52]: [ErrorReply: NOAUTH Authentication required.] this is the error i got

Comment (krishnaacharyaa) [reply backend/controllers/posts-controller.ts:52]: Can you please figure out why we require this, 
Have you setup redis locally in the said port given in the env? 

Once run redis locally, and see if you can figure out something from it, you would mostly like be able to do it!

File: backend/package.json (hunk 2 above)
Comment (krishnaacharyaa) [reply backend/package.json:3]: But module is better than commonjs, 

> we have used this is somewhere in our codebase i have checked but i don't know where we have used that
Can you figure out where and rectify by searching ctrl + shift +f   

If not only then we'll change, 
Let's not change because we are getting some error, instead lets see which is best commonjs or module? And then change our codebase

File: backend/tsconfig.json (hunk 3 above)
Comment (krishnaacharyaa) [reply backend/tsconfig.json:1]: commented code, please remove

File: backend/package.json (hunk 2 above)
Comment (Ameerjafar) [reply backend/package.json:3]: In the ts config also they the module is mentioned as common js @krishnaacharyaa

File: backend/tsconfig.json (hunk 3 above)
Comment (Ameerjafar) [reply backend/tsconfig.json:1]: done sir @krishnaacharyaa

File: backend/controllers/posts-controller.ts (hunk 1 above)
Comment (Ameerjafar) [reply backend/controllers/posts-controller.ts:52]: @krishnaacharyaa yes sir it is working well I just gave the correct port number.

**Comment by Ameerjafar** [comment]
@krishnaacharyaa

File: backend/controllers/posts-controller.ts (hunk 1 above)
Comment (Ameerjafar) [reply backend/controllers/posts-controller.ts]: @krishnaacharyaa it shows the error in the backend/dist/server.js file (Object.defineProperty(exports, "__esModule", { value: true });) exports is not defined in the es module scope.

**Comment by krishnaacharyaa** [comment]
Thoda late karpaaungaa reviews, sorry yaar, kaam bohot hei, don't mind 
I'll not be able to respond your messages in github/discord

File: backend/package.json (hunk 2 above)
Comment (Sukomal07) [reply backend/package.json:3]: @krishnaacharyaa Yeah, module is better than commonjs.

Comment (Sukomal07) [reply backend/package.json:3]: > In the type module we cannot use the exports or module.exports(we have used this is somewhere in our codebase i have checked but i don't know where we have used that. this is the reason i have useed type="commonjs" @krishnaacharyaa

Use export syntax instead of module.exports
example:
`
export const replaceStr = (str, char, replacer) => {
  const regex = new RegExp(char, "g")
  const replaced = str.replace(regex, replacer)
  return replaced
}
`

Comment (Ameerjafar) [reply backend/package.json:3]: @Sukomal07 This is the js code compiled by the ts compiler, In my ts code i did not use the module.exports.

File: backend/controllers/posts-controller.ts (hunk 1 above)
Comment (Ameerjafar) [reply backend/controllers/posts-controller.ts]: This problem is resolved @krishnaacharyaa

**Comment by Ameerjafar** [comment]
@Sukomal07

**Comment by Ameerjafar** [comment]
@krishnaacharyaa i have done if you time please review the pr sir. I have keep the type as module in the package.json file.

**Review by Sukomal07 — APPROVED** [review]

**Review by krishnaacharyaa — APPROVED** [review]
Thank you for the efforts @Ameerjafar, Much appreciated !
